pkg/engine/templating/resource_ref.go

```diff
@@ -24,17 +25,27 @@ func ResourceRef(ctx context.Context, c compilers.Compilers, obj *unstructured.U
 	temp.SetKind(kind)
 	temp.SetName(obj.GetName())
 	temp.SetNamespace(obj.GetNamespace())
-	temp.SetLabels(obj.GetLabels())
+	temp.SetLabels(getStringLabels(obj))
 	template := v1alpha1.NewProjection(temp.UnstructuredContent())
 	if merged, err := TemplateAndMerge(ctx, c, temp, bindings, template); err != nil {
 		return err
 	} else {
-		temp = merged
+		obj.Object = maps.Merge(obj.Object, merged.UnstructuredContent())
+	}
+	return nil
+}
+
+func getStringLabels(obj *unstructured.Unstructured) map[string]string {
+	if labels, ok, _ := unstructured.NestedFieldNoCopy(obj.UnstructuredContent(), "metadata", "labels"); ok {
+		if labelsMap, ok := labels.(map[string]interface{}); ok {
+			tempLabels := make(map[string]string)
+			for k, v := range labelsMap {
+				if stringValue, ok := v.(string); ok {
+					tempLabels[k] = stringValue
+				}
+			}
+			return tempLabels
+		}
```

Comment:
Do we need this or the merge line 33 is enough ?

Reply:
That, or a change in the implementation of `GetLabels()`. Or maybe there is some other helper function or simpler way to implement this. The 'problem' with the current `GetLabels()` is that it will return no labels at all when there is at least one label with a non-string value.